File: Java/src/com/dbc/algorithm/ReferOffer/ShuZuZhongShuZiChuXianDeCiShuIILcof.java
package com.dbc.algorithm.ReferOffer;

public class ShuZuZhongShuZiChuXianDeCiShuIILcof {
    public int singleNumber(int[] nums) {
        int[] counts = new int[32];
        for (int num : nums) {
            for (int i = 0; i < 32; i++) {
                counts[i] += num & 1;
                num >>= 1;
            }
        }
        int res = 0;
        for (int i = 31; i >= 0; i--) {
            res <<= 1;
            if (counts[i] % 3 == 1) {
                res += 1;
            }
        }
        return res;
    }
}

File: Python/refer-offer/shu_zu_zhong_shu_zi_chu_xian_de_ci_shu_ii_lcof.py
from typing import *

def singleNumber(nums: List[int]) -> int:
    count = [0] * 32
    for num in nums:
        for i in range(32):
            count[i] += num & 1
            num >>= 1

    res = 0
    for i in range(31, -1, -1):
        res <<= 1
        if count[i] % 3 == 1:
            res += 1
    return res

if __name__ == "__main__":
    print(singleNumber([9,1,7,9,7,9,7]))